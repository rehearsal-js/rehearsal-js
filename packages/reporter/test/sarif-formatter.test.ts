import { describe, expect, test } from 'vitest';

import { SarifFormatter } from '../src/formatters/index.js';
import {
  addArtifactData,
  addResultData,
  addRuleData,
  initialData,
} from './fixtures/sarif-formatter/index.js';
import type { Log } from 'sarif';

describe('SarifFormatter', () => {
  test('should set the correct version, $schema, and initial data', () => {
    expect(SarifFormatter.extension).toBe('.sarif');
    expect(SarifFormatter.getReport(initialData)).toMatchSnapshot();
  });

  test('should add a rule only if the rule does not exist in rules', () => {
    const log = JSON.parse(SarifFormatter.getReport(addRuleData)) as Log;

    expect(log.runs[0].tool.driver.rules).toMatchSnapshot();
  });

  test('should have the correct number of results in order, and each result should be assigned the correct property values', () => {
    const log = JSON.parse(SarifFormatter.getReport(addResultData)) as Log;

    expect(log.runs[0].results).toMatchSnapshot();
  });

  test('should add an artifact only if it does not exist, should merge roles and properties correctly', () => {
    const log = JSON.parse(SarifFormatter.getReport(addArtifactData)) as Log;

    expect(log.runs[0].artifacts).toMatchSnapshot();
  });
});
