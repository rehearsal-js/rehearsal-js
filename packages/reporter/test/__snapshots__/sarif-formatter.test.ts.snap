// Vitest Snapshot v1

exports[`Test sarif-formatter > should add a rule only if the rule does not exist in rules 1`] = `
"[
  {
    \\"id\\": \\"TS6133\\",
    \\"name\\": \\"TS6133\\",
    \\"shortDescription\\": {
      \\"text\\": \\"'resolve' is declared but its value is never read.\\"
    },
    \\"helpUri\\": \\"\\"
  },
  {
    \\"id\\": \\"TS2322\\",
    \\"name\\": \\"TS2322\\",
    \\"shortDescription\\": {
      \\"text\\": \\"Type 'string' is not assignable to type 'number'.\\"
    },
    \\"helpUri\\": \\"\\"
  },
  {
    \\"id\\": \\"TS2345\\",
    \\"name\\": \\"TS2345\\",
    \\"shortDescription\\": {
      \\"text\\": \\"Argument of type 'number' is not assignable to parameter of type 'string'.\\"
    },
    \\"helpUri\\": \\"\\"
  },
  {
    \\"id\\": \\"@typescript-eslint/no-explicit-any\\",
    \\"name\\": \\"@typescript-eslint/no-explicit-any\\",
    \\"shortDescription\\": {
      \\"text\\": \\"Unexpected any. Specify a different type.\\"
    },
    \\"helpUri\\": \\"\\"
  }
]"
`;

exports[`Test sarif-formatter > should add an artifact only if it does not exist, should merge roles and properties correctly 1`] = `
"[
  {
    \\"location\\": {
      \\"uri\\": \\"add-artifact-2.ts\\"
    }
  },
  {
    \\"location\\": {
      \\"uri\\": \\"add-artifact-4.ts\\"
    }
  }
]"
`;

exports[`Test sarif-formatter > should have the correct number of results in order, and each result should be assigned the correct property values 1`] = `
"[
  {
    \\"ruleId\\": \\"TS6133\\",
    \\"ruleIndex\\": 0,
    \\"level\\": \\"error\\",
    \\"kind\\": \\"fail\\",
    \\"message\\": {
      \\"text\\": \\"The declaration 'react' is never read or used. Remove the declaration or use it.\\"
    },
    \\"analysisTarget\\": {
      \\"uri\\": \\"add-result-1.ts\\"
    },
    \\"locations\\": [
      {
        \\"physicalLocation\\": {
          \\"artifactLocation\\": {
            \\"uri\\": \\"add-result-1.ts\\",
            \\"index\\": 0
          },
          \\"region\\": {
            \\"startLine\\": 1,
            \\"startColumn\\": 1,
            \\"endLine\\": 1,
            \\"endColumn\\": 20
          }
        }
      }
    ]
  },
  {
    \\"ruleId\\": \\"TS2322\\",
    \\"ruleIndex\\": 1,
    \\"level\\": \\"error\\",
    \\"kind\\": \\"fail\\",
    \\"message\\": {
      \\"text\\": \\"The variable 'a' has type 'number', but 'string' is assigned. Please convert 'string' to 'number' or change variable's type.\\"
    },
    \\"analysisTarget\\": {
      \\"uri\\": \\"add-result-1.ts\\"
    },
    \\"locations\\": [
      {
        \\"physicalLocation\\": {
          \\"artifactLocation\\": {
            \\"uri\\": \\"add-result-1.ts\\",
            \\"index\\": 0
          },
          \\"region\\": {
            \\"startLine\\": 3,
            \\"startColumn\\": 1,
            \\"endLine\\": 1,
            \\"endColumn\\": 5
          }
        }
      }
    ]
  },
  {
    \\"ruleId\\": \\"TS2345\\",
    \\"ruleIndex\\": 2,
    \\"level\\": \\"error\\",
    \\"kind\\": \\"fail\\",
    \\"message\\": {
      \\"text\\": \\"Argument of type '{0}' is not assignable to parameter of type 'string'. Consider verifying both types, using type assertion: '({} as string)', or using type guard: 'if ({} instanceof string) { ... }'.\\"
    },
    \\"analysisTarget\\": {
      \\"uri\\": \\"add-result-1.ts\\"
    },
    \\"locations\\": [
      {
        \\"physicalLocation\\": {
          \\"artifactLocation\\": {
            \\"uri\\": \\"add-result-1.ts\\",
            \\"index\\": 0
          },
          \\"region\\": {
            \\"startLine\\": 3,
            \\"startColumn\\": 10,
            \\"endLine\\": 10,
            \\"endColumn\\": 5
          }
        }
      }
    ]
  },
  {
    \\"ruleId\\": \\"@typescript-eslint/no-explicit-any\\",
    \\"ruleIndex\\": 3,
    \\"level\\": \\"error\\",
    \\"kind\\": \\"fail\\",
    \\"message\\": {
      \\"text\\": \\"Unexpected any. Specify a different type.\\"
    },
    \\"analysisTarget\\": {
      \\"uri\\": \\"add-rule-5.ts\\"
    },
    \\"locations\\": [
      {
        \\"physicalLocation\\": {
          \\"artifactLocation\\": {
            \\"uri\\": \\"add-rule-5.ts\\",
            \\"index\\": 1
          },
          \\"region\\": {
            \\"startLine\\": 7,
            \\"startColumn\\": 1,
            \\"endLine\\": 4,
            \\"endColumn\\": 7
          }
        }
      }
    ]
  }
]"
`;

exports[`Test sarif-formatter > should set the correct version, $schema, and initial data 1`] = `
"{
  \\"version\\": \\"2.1.0\\",
  \\"$schema\\": \\"http://json.schemastore.org/sarif-2.1.0-rtm.4\\",
  \\"runs\\": [
    {
      \\"tool\\": {
        \\"driver\\": {
          \\"name\\": \\"@rehearsal/reporter\\",
          \\"informationUri\\": \\"https://github.com/rehearsal-js/rehearsal-js\\",
          \\"rules\\": []
        }
      },
      \\"artifacts\\": [],
      \\"results\\": [],
      \\"automationDetails\\": {
        \\"description\\": {
          \\"text\\": \\"This is the run of @rehearsal/reporter on your product against TypeScript 4.7.4 at 9/16/2022, 20:16:50\\"
        }
      }
    }
  ]
}"
`;
